modules/core/src/test/java/org/apache/ignite/internal/processors/closure/ClusterMetricsOnCacheSelfTest.java

```diff
@@ -0,0 +1,172 @@
+
+            FifoEvictionPolicy plc = new FifoEvictionPolicy();
+            plc.setMaxMemorySize(MAX_VALS_AMOUNT * VAL_SIZE);
+            plc.setMaxSize(VAL_SIZE);
+
+            onHeapTieredCfg.setEvictionPolicy(plc);
+
+            cacheInit = true;
+
+            return cfg.setCacheConfiguration(offHeapValuesCfg, onHeapTieredCfg);
+        }
+
+        return cfg;
+    }
+
+    public void testAllocatedMemory() throws Exception {
+        final IgniteCache tiered = grid(0).getOrCreateCache(ON_HEAP_TIERED_NAME);
+        final IgniteCache values = grid(0).getOrCreateCache(OFF_HEAP_VALUE_NAME);
+
+        long prevTieredOffHeapSize = tiered.metrics().getOffHeapAllocatedSize();
+        long prevValuesOffHeapSize = values.metrics().getOffHeapAllocatedSize();
+
+        assertEquals(0, prevTieredOffHeapSize);
+        assertEquals(0, prevValuesOffHeapSize);
+
+        long prevClusterNonHeapMemoryUsed = grid(0).cluster().metrics().getNonHeapMemoryUsed();
+
+        fillCache(tiered, getTestTimeout());
+
+        assertTrue(tiered.metrics().getOffHeapAllocatedSize() > (MAX_VALS_AMOUNT - 5) * VAL_SIZE + prevTieredOffHeapSize);
+        assertEquals(prevValuesOffHeapSize, values.metrics().getOffHeapAllocatedSize());
+        // prevClusterNonHeapMemoryUsed = 69_277_368
+        // (MAX_VALS_AMOUNT - 5) * VAL_SIZE = 49_807_360
+        // tiered.metrics().getOffHeapAllocatedSize() = 51_012_531
+
+        assertTrue((MAX_VALS_AMOUNT - 5) * VAL_SIZE + prevClusterNonHeapMemoryUsed < grid(0).cluster().metrics().getNonHeapMemoryUsed());
+
+        prevClusterNonHeapMemoryUsed = grid(0).cluster().metrics().getNonHeapMemoryUsed();
+        prevTieredOffHeapSize = tiered.metrics().getOffHeapAllocatedSize();
+
+        fillCache(values, getTestTimeout());
+
+        assertTrue(values.metrics().getOffHeapAllocatedSize() > (MAX_VALS_AMOUNT - 5) * VAL_SIZE + prevValuesOffHeapSize);
+        assertEquals(prevTieredOffHeapSize, tiered.metrics().getOffHeapAllocatedSize());
+        assertTrue((MAX_VALS_AMOUNT - 5) * VAL_SIZE + prevClusterNonHeapMemoryUsed < grid(0).cluster().metrics().getNonHeapMemoryUsed());
+    }
+
+    /** Fill cache with values. */
+    private static void fillCache(final IgniteCache<Integer, Object> cache, long timeout) throws Exception{
+        final int THREAD_COUNT = 4;
+        final byte[] val = new byte[VAL_SIZE];
+        final AtomicInteger keyStart = new AtomicInteger(0);
+        final CountDownLatch latch = new CountDownLatch(THREAD_COUNT);
+
+        GridTestUtils.runMultiThreaded(new Callable<Void>() {
+            @Override public Void call() throws Exception {
+                final int start = keyStart.addAndGet(MAX_VALS_AMOUNT);
+
+                for (int i = start; i < start + MAX_VALS_AMOUNT; i++)
+                    cache.put(i, val);
+
+                latch.countDown();
+
+                return null;
+            }
+        }, THREAD_COUNT, "test");
+
+        latch.await(timeout, TimeUnit.SECONDS);
```

Comment:
There is no need to way while the cache is propagated. GridTestUtils.runMultiThreaded will block execution of your Thread until all the Threads created by this method are finished. See the source code of GridTestUtils.runMultiThreaded for more details.